Implement css property contain-intrinsic-size

cathiechen opened the pull request:
#### 7ed83ab2929f7a2be55fb40888fea7ff02909f56
<pre>
Implement css property contain-intrinsic-size
<a href="https://bugs.webkit.org/show_bug.cgi?id=238867">https://bugs.webkit.org/show_bug.cgi?id=238867</a>

Reviewed by Alan Bujtas.

This patch adds support for contain-intrinsic-size value &quot;none | &lt;length&gt;&quot;. The value &quot;auto &amp;&amp; &lt;length&gt;&quot;
is not supported, because per [1] it depends on &quot;content-visibility:auto&quot; which is not supported yet.
Per [2], the properties allow elements with size containment to specify an explicit intrinsic inner size,
which is determined in RenderBox. These values will affect the intrinsic width calculation in computeIntrinsicLogicalWidths.
As to intrinsic height, the value will be overridden at the point after layout children and before logical height calculation.
For grid layout, we need to use explicit intrinsic inner size to calculate auto repetition, if width/height is auto.

[1] <a href="https://github.com/w3c/csswg-drafts/issues/6308">https://github.com/w3c/csswg-drafts/issues/6308</a>
[2] <a href="https://www.w3.org/TR/css-sizing-4/#intrinsic-size-override">https://www.w3.org/TR/css-sizing-4/#intrinsic-size-override</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-009-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-029-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-030-expected.txt: The three failed cases are related to css scrollbar-gutter.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-031-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-032-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-logical-003-expected.txt:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::computeGridContainerIntrinsicSizes): Set explicitIntrinsicInnerLogicalSize to m_minContentSize and m_maxContentSize.
(WebCore::GridTrackSizingAlgorithm::run): Don&apos;t skip stretching tracks for size containment if there is explicitIntrinsicInnerLogicalSize.
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::updateLogicalHeight): Set explicitIntrinsicHeight to logicalHeight if it exists.
(WebCore::RenderBox::computeLogicalHeightWithoutLayout const): We should take account of explicitIntrinsicInnerLogicalHeight here.
(WebCore::RenderBox::explicitIntrinsicInnerWidth const):
(WebCore::RenderBox::explicitIntrinsicInnerHeight const):
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::explicitIntrinsicInnerLogicalWidth const):
(WebCore::RenderBox::explicitIntrinsicInnerLogicalHeight const):
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
(WebCore::RenderDeprecatedFlexibleBox::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderFileUploadControl.cpp:
(WebCore::RenderFileUploadControl::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutBlock): Should use explicitIntrinsicInnerHeigth as the definite size if exists.
(WebCore::RenderGrid::computeIntrinsicLogicalWidths const): Don&apos;t add gutterSize if it&apos;s explicitIntrinsicInnerLogicalSize.
(WebCore::RenderGrid::computeTrackSizesForIndefiniteSize const): Ditto.
(WebCore::RenderGrid::explicitIntrinsicInnerLogicalSize const): Ditto.
(WebCore::RenderGrid::computeAutoRepeatTracksCount const): Use explicitIntrinsicInnerLogicalSize as availableSize if the size of grid container is not specified.
(WebCore::RenderGrid::computeEmptyTracksForAutoRepeat const): The tracks of grid container with size containment are not empty if hasExplicitIntrinsicInnerLogicalSize.
* Source/WebCore/rendering/RenderGrid.h:
* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::RenderListBox::computeIntrinsicLogicalWidths const): Ditto.
(WebCore::RenderListBox::computeLogicalHeight const): Use explicitIntrinsicInnerLogicalHeight as estimated height not the sum of item height.
* Source/WebCore/rendering/RenderMenuList.cpp:
(RenderMenuList::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::computeIntrinsicRatioInformation const): Don&apos;t let size containment effect aspect ratio.
* Source/WebCore/rendering/RenderReplaced.h: Return explicitIntrinsicInnerSize if exits.
* Source/WebCore/rendering/RenderSlider.cpp:
(WebCore::RenderSlider::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderTextControl.cpp:
(WebCore::RenderTextControl::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderVideo.cpp:
(WebCore::RenderVideo::calculateIntrinsicSize):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareNonInheritedDataChangeRequiresLayout): The changes of contain-intrinsic-size should require layout.

Canonical link: <a href="https://commits.webkit.org/255971@main">https://commits.webkit.org/255971@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5b0828900abb00fe80b37899e52d427b18a495b7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94190 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3382 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/24717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103813 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164144 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98179 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3401 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31596 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86502 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99834 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99845 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/2440 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/80608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29461 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84364 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/24717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72410 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37978 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/24717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35863 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/24717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4129 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39738 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/80608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41680 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/24717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -790,6 +804,9 @@ override;
     void applyTopLeftLocationOffsetWithFlipping(LayoutPoint&) const;
 
+    bool hasExplicitIntrinsicInnerWidth() const;
+    bool hasExplicitIntrinsicInnerHeight() const;
```
It looks like these two methods are not needed, assuming they will only be used internally in RenderBox.cpp.

cathiechen replied:
Done, thanks!

rwlbuis commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -98,8 +98,12 @@ void RenderFlexibleBox::computeIntrinsicLogicalWidths(LayoutUnit& minLogicalWidt
 
-    auto shouldIgnoreFlexItemContentForLogicalWidth = shouldApplySizeOrInlineSizeContainment();
-    if (shouldIgnoreFlexItemContentForLogicalWidth) {
+    bool isSizeContainment = shouldApplySizeContainment();
```
Better here is needsSizeContainment.

cathiechen replied:
Done

rwlbuis commented on LayoutTests/TestExpectations:
```diff
@@ -4729,34 +4729,7 @@ webkit.org/b/214463 imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/
 webkit.org/b/214463 imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/floats-aspect-ratio-001.html [ ImageOnlyFailure ]
 
 # contain-intrinsic-size
```
Probably this comment can be removed too.

cathiechen replied:
Done

rwlbuis commented on Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
```diff
@@ -202,6 +202,8 @@ void RenderDeprecatedFlexibleBox::computeIntrinsicLogicalWidths(LayoutUnit& minL
 
     if (shouldApplySizeContainment()) {
+        if (auto width = explicitIntrinsicInnerLogicalWidth())
+            minLogicalWidth = maxLogicalWidth = width.value();
```
These two lines happen a lot, could we have a shared method?

rwlbuis replied:
Ok, I misread, I thought we were setting member vars here. I don't think we can improve much here.

rwlbuis commented on Source/WebCore/rendering/RenderVideo.cpp:
```diff
@@ -117,8 +117,14 @@ bool RenderVideo::updateIntrinsicSize()
+            size.setWidth(width.value());
+        if (auto height = explicitIntrinsicInnerHeight())
+            size.setHeight(height.value());
+        return size;
```
Is this needed since we have RenderReplaced:: intrinsicSize? If yes, can we share the code?

cathiechen replied:
I'm not quite sure what does this mean?

rwlbuis replied:
RenderVideo is a RenderReplaced, so why is the RenderReplaced:: intrinsicSize change not enough?

cathiechen replied:
OK, makes sense. I don't quite remember, is this to solve some issue or not.  Let me try to remove it.

rwlbuis replied:
It seems to work! Can you adjust the commit message to explain this better?:
 (WebCore::RenderVideo::calculateIntrinsicSize): Ditto.
I think github allows to edit the commit message by hand...

cathiechen replied:
Sure

cathiechen commented on LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-030-expected.txt:
```diff
@@ -1,62 +1,32 @@
-<div class="test small-contents overflow-auto cis-height" data-expected-client-width="0" data-expected-client-height="50" data-expected-scroll-width="40" data-expected-scroll-height="50"></div>
-clientHeight expected 50 but got 0
+PASS .test 2
 FAIL .test 3 assert_equals:
```
The cases 3, 15 and 16 use `scrollbar-gutter: stable;` to turn on the scrollbar. `scrollbar-gutter` is not supported on WebKit yet, so these cases are failing now.
Left a comment on bug: 167335

cathiechen commented on LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-032-expected.txt:
```diff
@@ -20,11 +20,11 @@ FAIL .test 6 assert_equals:
 clientWidth expected 0 but got 34
 FAIL .test 7 assert_equals:
 <select class="test ciw-100 cih-none" data-expected-client-width="100" data-expected-client-height="18"></select>
```
The failures in this test are related to https://bugs.webkit.org/show_bug.cgi?id=246338

cathiechen commented on LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-logical-003-expected.txt:
```diff
@@ -1,128 +1,65 @@
+PASS .test 23
+PASS .test 24
 FAIL .test 25 assert_equals:
 <canvas class="test cis-none" width="40" height="20" data-expected-client-width="0" data-expected-client-height="0"></canvas>
```
These are related to https://bugs.webkit.org/show_bug.cgi?id=246283

cathiechen commented on LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-003-expected.txt:
```diff
@@ -1,3 +1,3 @@
 
-FAIL contain-intrinsic-size: auto assert_equals: expected 1 but got 0
+FAIL contain-intrinsic-size: auto assert_equals: expected 50 but got 1
```
`contain-intrinsic-size: auto && <length>` is not supported yet, it depends on `content-visibility`. https://bugs.webkit.org/show_bug.cgi?id=239450

alanbaradlay commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -780,6 +780,13 @@ double GridTrackSizingAlgorithm::findFrUnitSize(const GridSpan& tracksSpan, Layo
 {
+    if (m_direction == ForColumns && m_strategy->isComputingSizeContainment()) {
+        if (auto size = m_renderGrid->explicitIntrinsicInnerLogicalSize(m_direction)) {
+            m_minContentSize = m_maxContentSize = size.value();
```
use the WebKit style here, please (and at all the other places)
m_minContentSize = size.value();
m_maxContentSize = size.value();

cathiechen replied:
Done, thanks!

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5555,4 +5563,26 @@ bool RenderBox::shouldIgnoreMinMaxSizes() const
+
+    auto width = style().containIntrinsicWidth();
+    ASSERT(width.has_value());
+    return std::optional<LayoutUnit>(width.value().value());
```
you could just do (same below) 
return { width->value() }

cathiechen replied:
Ah, yes, it looks better!
But compiler complains: `error: chosen constructor is explicit in copy-initialization`.
So I use `return std::optional<LayoutUnit>{ width->value() };` in the new patch

alanbaradlay reviewed (approved):
While I can't tell if this covers all the cases (and whether it's missing some pieces), we should land it and watch for the fallout.

cathiechen committed:
Implement css property contain-intrinsic-size
https://bugs.webkit.org/show_bug.cgi?id=238867

Reviewed by Alan Bujtas.

This patch adds support for contain-intrinsic-size value "none | <length>". The value "auto && <length>"
is not supported, because per [1] it depends on "content-visibility:auto" which is not supported yet.
Per [2], the properties allow elements with size containment to specify an explicit intrinsic inner size,
which is determined in RenderBox. These values will affect the intrinsic width calculation in computeIntrinsicLogicalWidths.
As to intrinsic height, the value will be overridden at the point after layout children and before logical height calculation.
For grid layout, we need to use explicit intrinsic inner size to calculate auto repetition, if width/height is auto.

[1] https://github.com/w3c/csswg-drafts/issues/6308
[2] https://www.w3.org/TR/css-sizing-4/#intrinsic-size-override

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-009-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-029-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-030-expected.txt: The three failed cases are related to css scrollbar-gutter.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-031-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-032-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-logical-003-expected.txt:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::computeGridContainerIntrinsicSizes): Set explicitIntrinsicInnerLogicalSize to m_minContentSize and m_maxContentSize.
(WebCore::GridTrackSizingAlgorithm::run): Don't skip stretching tracks for size containment if there is explicitIntrinsicInnerLogicalSize.
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::updateLogicalHeight): Set explicitIntrinsicHeight to logicalHeight if it exists.
(WebCore::RenderBox::computeLogicalHeightWithoutLayout const): We should take account of explicitIntrinsicInnerLogicalHeight here.
(WebCore::RenderBox::explicitIntrinsicInnerWidth const):
(WebCore::RenderBox::explicitIntrinsicInnerHeight const):
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::explicitIntrinsicInnerLogicalWidth const):
(WebCore::RenderBox::explicitIntrinsicInnerLogicalHeight const):
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
(WebCore::RenderDeprecatedFlexibleBox::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderFileUploadControl.cpp:
(WebCore::RenderFileUploadControl::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutBlock): Should use explicitIntrinsicInnerHeigth as the definite size if exists.
(WebCore::RenderGrid::computeIntrinsicLogicalWidths const): Don't add gutterSize if it's explicitIntrinsicInnerLogicalSize.
(WebCore::RenderGrid::computeTrackSizesForIndefiniteSize const): Ditto.
(WebCore::RenderGrid::explicitIntrinsicInnerLogicalSize const): Ditto.
(WebCore::RenderGrid::computeAutoRepeatTracksCount const): Use explicitIntrinsicInnerLogicalSize as availableSize if the size of grid container is not specified.
(WebCore::RenderGrid::computeEmptyTracksForAutoRepeat const): The tracks of grid container with size containment are not empty if hasExplicitIntrinsicInnerLogicalSize.
* Source/WebCore/rendering/RenderGrid.h:
* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::RenderListBox::computeIntrinsicLogicalWidths const): Ditto.
(WebCore::RenderListBox::computeLogicalHeight const): Use explicitIntrinsicInnerLogicalHeight as estimated height not the sum of item height.
* Source/WebCore/rendering/RenderMenuList.cpp:
(RenderMenuList::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::computeIntrinsicRatioInformation const): Don't let size containment effect aspect ratio.
* Source/WebCore/rendering/RenderReplaced.h: Return explicitIntrinsicInnerSize if exits.
* Source/WebCore/rendering/RenderSlider.cpp:
(WebCore::RenderSlider::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderTextControl.cpp:
(WebCore::RenderTextControl::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderVideo.cpp:
(WebCore::RenderVideo::calculateIntrinsicSize):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareNonInheritedDataChangeRequiresLayout): The changes of contain-intrinsic-size should require layout.

Canonical link: https://commits.webkit.org/255971@main
